File: spec/examples/typescript/generate-index.ts
#!/usr/bin/env -S node --experimental-strip-types
// Minimal reference generator proving the index format needs no library
// beyond a JSON encoder — deliberately independent of this repo's own
// @csf/* packages (no shared code with the real indexer). See
// docs/02-index-format.md and docs/20-tech-stack.md. Usage:
//   node --experimental-strip-types generate-index.ts documents.json out/
//   (or: npx tsx generate-index.ts documents.json out/)
import { createHash } from "node:crypto";
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";

interface SourceDoc {
  id: number;
  url: string;
  title: string;
  body: string;
}

interface FieldPosting {
  tf: number;
  pos: number[];
  len: number;
}

interface Posting {
  doc: number;
  fields: Record<string, FieldPosting>;
}

interface TermEntry {
  df: number;
  postings: Posting[];
}

function tokenize(text: string): string[] {
  const stripped = text.replace(/<[^>]+>/g, " ");
  return stripped.toLowerCase().match(/[a-z0-9]+/g) ?? [];
}

function writeJson(outDir: string, relPath: string, data: unknown): string {
  const json = JSON.stringify(data);
  const hash = createHash("sha256").update(json).digest("hex").slice(0, 8);
  const hashedRelPath = relPath.replace(/\.json$/, `.${hash}.json`);
  const absPath = join(outDir, hashedRelPath);
  mkdirSync(dirname(absPath), { recursive: true });
  writeFileSync(absPath, json);
  return hashedRelPath;
}

const [inputFile, outDir] = process.argv.slice(2) as [string, string];
const docs: SourceDoc[] = JSON.parse(readFileSync(inputFile, "utf8"));

const terms: Record<string, TermEntry> = {};
const docStore: Record<
  string,
  { url: string; fields: Record<string, string> }
> = {};
let titleLenSum = 0;
let bodyLenSum = 0;

for (const doc of docs) {
  const titleTokens = tokenize(doc.title);
  const bodyTokens = tokenize(doc.body);
  titleLenSum += titleTokens.length;
  bodyLenSum += bodyTokens.length;

  const fieldTokens: [string, string[]][] = [
    ["title", titleTokens],
    ["body", bodyTokens],
  ];
  for (const [field, tokens] of fieldTokens) {
    const positionsByTerm: Record<string, number[]> = {};
    tokens.forEach((term, pos) => {
      if (!positionsByTerm[term]) positionsByTerm[term] = [];
      positionsByTerm[term].push(pos);
    });
    for (const [term, positions] of Object.entries(positionsByTerm)) {
      terms[term] ??= { df: 0, postings: [] };
      let posting = terms[term].postings.find((p) => p.doc === doc.id);
      if (!posting) {
        posting = { doc: doc.id, fields: {} };
        terms[term].postings.push(posting);
        terms[term].df++;
      }
      posting.fields[field] = {
        tf: positions.length,
        pos: positions,
        len: tokens.length,
      };
    }
  }

  docStore[String(doc.id)] = {
    url: doc.url,
    fields: { title: doc.title, excerpt: doc.body.slice(0, 200) },
  };
}

const termsFile = writeJson(outDir, "terms/en/all.json", terms);
const docsFile = writeJson(outDir, "docs/0.json", docStore);
const ids = docs.map((d) => d.id);

const manifest = {
  version: 1,
  buildId: new Date().toISOString(),
  format: "json",
  languages: ["en"],
  defaultLanguage: "en",
  fields: {
    title: { boost: 1.0, stored: true },
    body: { boost: 1.0, stored: false },
  },
  docCount: docs.length,
  avgFieldLength: {
    title: titleLenSum / docs.length,
    body: bodyLenSum / docs.length,
  },
  shards: {
    terms: [
      {
        lang: "en",
        prefix: "all",
        file: termsFile,
        termCount: Object.keys(terms).length,
      },
    ],
    docs: [
      {
        shard: 0,
        file: docsFile,
        idRange: [Math.min(...ids), Math.max(...ids)],
      },
    ],
  },
};

writeFileSync(join(outDir, "manifest.json"), JSON.stringify(manifest));
console.log(`indexed ${docs.length} document(s) -> ${outDir}`);

File: spec/examples/python/generate_index.py
#!/usr/bin/env python3
"""
Minimal reference generator proving the index format needs no library
beyond the standard library (json, hashlib, re) -- deliberately
independent of this repo's own TypeScript packages, sharing no code
with the real indexer. See docs/concepts/index-format.md and
docs/concepts/architecture.md.

Usage: python3 generate_index.py documents.json out/
"""
import hashlib
import json
import os
import re
import sys
from datetime import datetime, timezone


def tokenize(text):
    stripped = re.sub(r"<[^>]+>", " ", text)
    return re.findall(r"[a-z0-9]+", stripped.lower())


def write_json(out_dir, rel_path, data):
    content = json.dumps(data)
    digest = hashlib.sha256(content.encode()).hexdigest()[:8]
    hashed_rel = re.sub(r"\.json$", f".{digest}.json", rel_path)
    abs_path = os.path.join(out_dir, hashed_rel)
    os.makedirs(os.path.dirname(abs_path), exist_ok=True)
    with open(abs_path, "w") as f:
        f.write(content)
    return hashed_rel


def main():
    input_file, out_dir = sys.argv[1], sys.argv[2]
    with open(input_file) as f:
        docs = json.load(f)

    terms = {}
    doc_store = {}
    title_len_sum = 0
    body_len_sum = 0

    for doc in docs:
        title_tokens = tokenize(doc["title"])
        body_tokens = tokenize(doc["body"])
        title_len_sum += len(title_tokens)
        body_len_sum += len(body_tokens)

        for field, tokens in (("title", title_tokens), ("body", body_tokens)):
            positions_by_term = {}
            for pos, term in enumerate(tokens):
                positions_by_term.setdefault(term, []).append(pos)
            for term, positions in positions_by_term.items():
                entry = terms.setdefault(term, {"df": 0, "postings": []})
                posting = next(
                    (p for p in entry["postings"] if p["doc"] == doc["id"]), None
                )
                if posting is None:
                    posting = {"doc": doc["id"], "fields": {}}
                    entry["postings"].append(posting)
                    entry["df"] += 1
                posting["fields"][field] = {
                    "tf": len(positions),
                    "pos": positions,
                    "len": len(tokens),
                }

        doc_store[str(doc["id"])] = {
            "url": doc["url"],
            "fields": {"title": doc["title"], "excerpt": doc["body"][:200]},
        }

    terms_file = write_json(out_dir, "terms/en/all.json", terms)
    docs_file = write_json(out_dir, "docs/0.json", doc_store)
    ids = [d["id"] for d in docs]

    manifest = {
        "version": 2,
        "buildId": datetime.now(timezone.utc).isoformat(),
        "languages": ["en"],
        "defaultLanguage": "en",
        "fields": {
            "title": {"boost": 1.0, "stored": True},
            "body": {"boost": 1.0, "stored": False},
        },
        "docCount": {"en": len(docs)},
        "avgFieldLength": {
            "en": {
                "title": title_len_sum / len(docs),
                "body": body_len_sum / len(docs),
            }
        },
        "shards": {
            "terms": [
                {
                    "lang": "en",
                    "prefix": "all",
                    "file": terms_file,
                    "termCount": len(terms),
                }
            ],
            "docs": [
                {"shard": 0, "file": docs_file, "idRange": [min(ids), max(ids)]}
            ],
        },
    }

    with open(os.path.join(out_dir, "manifest.json"), "w") as f:
        json.dump(manifest, f)

    print(f"indexed {len(docs)} document(s) -> {out_dir}")


if __name__ == "__main__":
    main()
